File: iosApp/iosApp/BackendFlatVacantRequestRepository.swift
import Foundation
import FirebaseFirestore

struct FlatVacantRequestItem: Identifiable, Hashable {
    let id: String
    let occupantId: String
    let occupantName: String
    let flatId: String
    let flatNumber: String
    let reason: String
    let status: String
    let adminRemarks: String
    let dateCreated: Date
}

class BackendFlatVacantRequestRepository {
    private let db = Firestore.firestore()
    private let collection = "FlatVacantRequests"

    func submitRequest(
        occupantId: String,
        occupantName: String,
        flatId: String,
        flatNumber: String,
        reason: String
    ) async throws -> String {
        let data: [String: Any] = [
            "OccupantId": occupantId,
            "OccupantName": occupantName,
            "flatId": flatId,
            "FlatNumber": flatNumber,
            "Reason": reason,
            "Status": "PENDING",
            "AdminRemarks": "",
            "DateCreated": Timestamp(date: Date())
        ]
        let ref = try await db.collection(collection).addDocument(data: data)
        return ref.documentID
    }

    func observeByOccupant(
        occupantId: String,
        onChange: @escaping ([FlatVacantRequestItem]) -> Void
    ) -> ListenerRegistration {
        return db.collection(collection)
            .whereField("OccupantId", isEqualTo: occupantId)
            .order(by: "DateCreated", descending: true)
            .addSnapshotListener { snapshot, _ in
                guard let snapshot = snapshot else { onChange([]); return }
                let items: [FlatVacantRequestItem] = snapshot.documents.map { doc in
                    let data = doc.data()
                    return FlatVacantRequestItem(
                        id: doc.documentID,
                        occupantId: data["OccupantId"] as? String ?? "",
                        occupantName: data["OccupantName"] as? String ?? "",
                        flatId: data["flatId"] as? String ?? "",
                        flatNumber: data["FlatNumber"] as? String ?? "",
                        reason: data["Reason"] as? String ?? "",
                        status: data["Status"] as? String ?? "PENDING",
                        adminRemarks: data["AdminRemarks"] as? String ?? "",
                        dateCreated: (data["DateCreated"] as? Timestamp)?.dateValue() ?? Date()
                    )
                }
                onChange(items)
            }
    }
}

File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/repository/BackendFlatVacantRequestRepository.kt
package com.humblesolutions.indsphinx.repository

import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.humblesolutions.indsphinx.model.FlatVacantRequest
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import java.util.Date

class BackendFlatVacantRequestRepository : FlatVacantRequestRepository {
    private val db = FirebaseFirestore.getInstance()
    private val collection = "FlatVacantRequests"

    override suspend fun submitRequest(request: FlatVacantRequest): String {
        val data = hashMapOf(
            "OccupantId" to request.occupantId,
            "OccupantName" to request.occupantName,
            "flatId" to request.flatId,
            "FlatNumber" to request.flatNumber,
            "Reason" to request.reason,
            "Status" to "PENDING",
            "AdminRemarks" to "",
            "DateCreated" to Timestamp(Date()),
        )
        val doc = db.collection(collection).add(data).await()
        return doc.id
    }

    override fun observeByOccupant(occupantId: String): Flow<List<FlatVacantRequest>> = callbackFlow {
        val registration = db.collection(collection)
            .whereEqualTo("OccupantId", occupantId)
            .orderBy("DateCreated", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, error ->
                if (error != null || snapshot == null) { trySend(emptyList()); return@addSnapshotListener }
                val items = snapshot.documents.map { doc ->
                    FlatVacantRequest(
                        id = doc.id,
                        occupantId = doc.getString("OccupantId") ?: "",
                        occupantName = doc.getString("OccupantName") ?: "",
                        flatId = doc.getString("flatId") ?: "",
                        flatNumber = doc.getString("FlatNumber") ?: "",
                        reason = doc.getString("Reason") ?: "",
                        status = doc.getString("Status") ?: "PENDING",
                        adminRemarks = doc.getString("AdminRemarks") ?: "",
                        dateCreated = (doc.get("DateCreated") as? Timestamp)?.toDate()?.time ?: 0L,
                    )
                }
                trySend(items)
            }
        awaitClose { registration.remove() }
    }
}
